src/python/pants/engine/rules.py

```diff
@@ -64,10 +83,21 @@ def resolve_type(name):
         gets.update(Get(resolve_type(p), resolve_type(s)) for p, s in rule_visitor.gets)
 
     func._rule = TaskRule(output_type, input_selectors, func, input_gets=list(gets))
+    func.output_type = output_type
+    func.goal = for_goal
     return func
   return wrapper
 
 
+def rule(output_type, input_selectors):
+  return _make_rule(output_type, input_selectors)
+
+
+def console_rule(goal_name, input_selectors):
+  output_type = GoalProduct.for_name(goal_name)
```

Comment:
If `Console` were a product type, inverting the moon-landing flow, alot of future goals for console handling might fall out a bit more naturally.

A sketch - here `IO` is installed as a singleton that is opaque with no methods to take advantage of in user code. In order for user rule code to create a `Console` (or `InteractiveConsole` (stdin)) or `Logger`, they must pass in an `IO`. Behind the scenes it would probably carry a `Clock`.
```python
@rule(Console, [Select(IO), Select(BuildFileAddresses)], goal='list')
def fast_list(io, addresses):
  """A fast variant of `./pants list` with a reduced feature set."""
  console = Console(io)
  for address in addresses.dependencies:
    console.print_stdout(address.spec)
  return console
```

So now a `@console_rule` is just a `@rule` that produces a `Console` and has a user-facing CLI name (`goal='list'`). The console would have no handle to real io, it would just buffer data. The engine could then validate and aggregate and/or serialize CLI `@rule`s and `Console` products, actually driving the `Console` with one real console.

If the native engine were modified slightly to handle rules returning a generator - ie: feed generator rules until `StopIteration` instead of assuming a single yielded value, this could be:
```python
@rule(Stream(Console), [Select(IO), Select(BuildFileAddresses)], goal='list')
def fast_list(clock, addresses):
  """A fast variant of `./pants list` with a reduced feature set."""
  console = Console(io)
  for address in addresses.dependencies:
    yield console.print_stdout(address.spec)
```

Where `Console.print_stdout` returned a `Console`.

Reply:
Hm... I think that from my perspective, `Select(IO)` is approximately what `Select(Console)` does currently: ie, represents exclusive access to a console.

I'm not sure I see the advantage of the console being an output product when the other guarantees need to be met (exclusive, foreground access). Even in a world where we added back streaming of the console product `Stream(Console)`, the API would seem to be unnecessarily restrictive given that we know it "owns" the console while running.

Reply:
yeah, during design discussions I had initially proposed the idea of a `Console` as buffered virtual console in a similar vein to what you're suggesting (except without the Console/IO split - just wholesale copies) - but we'd decided against that in favor of more direct, exclusive access to accommodate a wider set of rule types to start.

I think for now, I agree with Stu that these are approximately equivalent - but I definitely see this iteratively evolving way beyond this current ~throwaway impl.